src/server/frontend_wayland/frame_callback_executor.h

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright © 2021 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: William Wold <william.wold@canonical.com>
+ */
+
+#ifndef MIR_FRONTEND_FRAME_CALLBACK_EXECUTOR_H
+#define MIR_FRONTEND_FRAME_CALLBACK_EXECUTOR_H
+
+#include <mir/executor.h>
+
+#include <vector>
+#include <mutex>
+#include <memory>
+
+namespace mir
+{
+namespace time
+{
+class Alarm;
+class AlarmFactory;
+}
+
+namespace frontend
+{
+
+/// Runs frame callbacks that do not have a buffer to be attached to. Is threadsafe. Given callbacks are run on the
+/// main loop thread, the wayland executor is NOT automatically used.
+class FrameCallbackExecutor : public Executor
```

Comment:
I think something is wrong here, probably the name.

If we really wanted to constrain the type of work to frame callbacks, then we should not derive from `Executor` (which takes any type of work) but have a `spawn` function that only accepts `WlSurface` objects on which it calls `send_frame_callbacks()`.

Instead, it should be named by the service it provides, not the expected usage. `VsyncExecutor` is clearly a lie, but how about `CompositionSyncExecutor`, or if "frame" is interpreted more broadly `FrameExecutor`?

Reply:
I did struggle a bit with the name. I'm going with `FrameExecutor`, since I think that is currently the most accurate.